fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingCommon.java

```diff
@@ -0,0 +1,176 @@
+import java.util.UUID;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.Function;
+
+/**
+ * Shared structure to be used (only) by the OnlineIndexer* modules.
+ */
+@API(API.Status.INTERNAL)
+public class IndexingCommon {
+    private final UUID uuid = UUID.randomUUID();
+
+    @Nonnull private final FDBDatabaseRunner runner;
+    @Nullable private SynchronizedSessionRunner synchronizedSessionRunner = null;
+
+    @Nonnull private final FDBRecordStore.Builder recordStoreBuilder;
+    @Nonnull private final Index index;
+    @Nonnull private final AtomicLong totalRecordsScanned;
+
+    private final boolean useSynchronizedSession;
+    private final boolean syntheticIndex;
+    private final boolean trackProgress;
+    private final long leaseLengthMillis;
+
+
+    @Nonnull private final OnlineIndexer.IndexStatePrecondition indexStatePrecondition;
+    @Nonnull public OnlineIndexer.Config config; // this item may be modified on the fly
+    @Nullable private final Function<OnlineIndexer.Config, OnlineIndexer.Config> configLoader;
+    private int configLoaderInvocationCount = 0;
+
+    @Nonnull public Collection<RecordType> recordTypes;
+
+    /**
+     * Constant indicating that there should be no limit to some usually limited operation.
+     */
+    public static final int UNLIMITED = Integer.MAX_VALUE;
+
+    IndexingCommon(@Nonnull FDBDatabaseRunner runner,
+                   @Nonnull FDBRecordStore.Builder recordStoreBuilder,
+                   @Nonnull Index index, @Nonnull Collection<RecordType> recordTypes,
+                   @Nonnull Function<OnlineIndexer.Config, OnlineIndexer.Config> configLoader, @Nonnull OnlineIndexer.Config config,
+                   boolean syntheticIndex,
+                   @Nonnull OnlineIndexer.IndexStatePrecondition indexStatePrecondition,
+                   boolean trackProgress,
+                   boolean useSynchronizedSession,
+                   long leaseLengthMillis ) {
+        this.useSynchronizedSession = useSynchronizedSession;
+        this.runner = runner;
+        this.index = index;
+        this.recordTypes = recordTypes;
+        this.configLoader = configLoader;
+        this.config = config;
+        this.syntheticIndex = syntheticIndex;
+        this.indexStatePrecondition = indexStatePrecondition;
+        this.trackProgress = trackProgress;
+        this.recordStoreBuilder = recordStoreBuilder;
+        this.leaseLengthMillis = leaseLengthMillis;
+
+        this.totalRecordsScanned = new AtomicLong(0);
+    }
+
+    public UUID getUuid() {
+        return uuid;
+    }
+
+    public boolean isUseSynchronizedSession() {
+        return useSynchronizedSession;
+    }
+
+    @Nonnull
+    public FDBDatabaseRunner getRunner() {
+        return synchronizedSessionRunner == null ? runner : synchronizedSessionRunner;
+    }
+
+    @Nonnull
+    public Index getIndex() {
+        return index;
+    }
+
+    public boolean isSyntheticIndex() {
```

Comment:
I've seen "synthetic" pop up a few times in this PR, but what is a synthetic index, or a synthetic record type?

Reply:
Yeah, they're indexes on synthetic record types (added to support join indexes; see: #306)